### PR TITLE
[Caffe Frontend] introduce caffe frontend for tvm

### PR DESCRIPTION
**Pull request by fernchen**
I mentioned a PR before, see https://github.com/apache/incubator-tvm/pull/6000, but because of my mistake, I merged the master branch. In order to correct this mistake, I propose this pr.

Please forgive me for my mistakes and help to review this pr, thanks you very much!
@tqchen @FrozenGene @siju-samuel 

### Background & Motivation

[Caffe](<https://caffe.berkeleyvision.org/>) is a deep learning framework made with expression, speed, and modularity in mind. Because of its simplicity, good scalability, fast and other characteristics, it is favored by many people. According to [riselab]( https://rise.cs.berkeley.edu/) who makes statistics of papers collected in arxiv.org, Caffe is ranked in the top four in the deep learning framework, which shows to some extent that Caffe's user base is still large, please refer to [blog](<https://www.oreilly.com/content/one-simple-graphic-researchers-love-pytorch-and-tensorflow/>). In addition, according to our company's research on the market, the demand for Caffe in the production environment is still strong, and many models based on Caffe need to be deployed. For example, existing deployment frameworks, such as [MNN](<https://github.com/alibaba/MNN>), [NCNN](<https://github.com/Tencent/ncnn>), [MACE](<https://github.com/XiaoMi/mace>), etc., directly support the deployment of Caffe.

TVM only supports caffe2 at present, and the difference between Caffe and caffe2 is quite large. At present, there are two ways to deploy Caffe model in TVM: one is to convert Caffe model to Tensorflow or Pytorch model, the other is to convert Caffe model to onnx and then to relay IR. The two methods are essentially the same. They are all indirectly converted to relay IR through the third-party transformation. However, the problem is that some ops will fail in the process of model transformation, and even the result of transfer out may be different.

Based on the above situation, we decided to open our Caffe frontend codes, hoping to enrich the use scenarios of TVM.

### Implementation Approach

The whole process of Caffe front end importing model is divided into:

1. Read Model：The model graph and related parameters are read through the protobuffer API of Caffe;
2. Rebuild Graph：Traverse the graph, then replace the top of the in-place layer with the name of the layer, and update all related layers at the same time;
3. Model Conversion：Read the parameters of each layer and convert them into corresponding TVM OP and parameters;
4. Layer Fusion：fuse batchnorm and scale layers;
5. Convert to Relay IR：It mainly includes its module, params and the real name of the output layer。

Finally, we can import the Caffe model as follows:

```Python
from google.protobuf import text_format
from caffe.proto import caffe_pb2 as pb

init_net = pb.NetParameter()
predict_net = pb.NetParameter()

# load model
with open(proto_file, 'r') as f:
    text_format.Merge(f.read(), predict_net)
# load blob
with open(blob_file, 'rb') as f:
    init_net.ParseFromString(f.read())

shape_dict = {'data': [1,3,224,224]}
dtype_dict = {'data': 'float32'}

mod, params = relay.frontend.from_caffe(init_net, predict_net, shape_dict, dtype_dict)
```

### Work Done

All of the things that we have done are listed as following:

#### 1. List of supported Ops

- [x] BatchNorm
- [x] Concat
- [x] Convolution
- [x] Crop
- [x] Deconvolution
- [x] Dropout
- [x] Eltwise
- [x] Flatten
- [x] InnerProduct
- [x] Input
- [x] LRN
- [ ] Normalize
- [ ] Permute
- [x] Pooling
- [x] PReLU
- [ ] PriorBox
- [ ] proposal
- [ ] Python
- [x] ReLU
- [x] Reshape
- [ ] Resize
- [ ] ROIPooling
- [x] Scale
- [x] Sigmoid
- [x] Slice
- [x] Softmax
- [x] TanH
- [ ] Upsample

#### 2. List of supported complete models

- [x] Alexnet
- [x] Resnet50
- [x] Mobilenetv1
- [x] Mobilenetv2
- [x] Inceptionv1
- [x] Inceptionv3
- [x] Inceptionv4
- [x] Vgg16
- [x] Squeezenetv1
- [ ] SSDMobilenetv1
- [ ] SSDMobilenetv2
- [ ] YOLOv3
- [ ] ENet

### TODO

- [ ] Support more ops and more complete models.

According to the above implementation scheme, based on the front-end framework we built, you can add any new op, you only need to: firstly, add a method in the operatorconverter class, which needs to include your extraction of the layer parameters and the logic of conversion to TVM OP, secondly, register the method to convert_ map.

### CODE REVIEW

File: tutorials/frontend/from_caffe.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+# Create a runtime executor module
+ctx = tvm.cpu(0)
+m = graph_runtime.GraphModule(lib['default'](ctx))
+
+# Feed input data
+m.set_input('data', tvm.nd.array(image_data))
+
+# Feed related params
+m.set_input(**params)
```
Comment (FrozenGene) [inline tutorials/frontend/from_caffe.py:117]: we don't need this in the new model based interface. Please remove this line.

File: python/tvm/relay/frontend/caffe.py [hunk 2]
```diff
@@ -0,0 +1,849 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# pylint: disable=invalid-name, unused-argument, too-many-lines, import-outside-toplevel
+# pylint: disable=no-else-return, no-else-continue
+"""Caffe frontend."""
+from __future__ import absolute_import as _abs
```
[the remark this answers is not in the document]
Comment (fernchen) [reply python/tvm/relay/frontend/caffe.py:21]: I will delete it later, thanks.

File: python/tvm/relay/frontend/caffe.py [hunk 3]
```diff
@@ -0,0 +1,849 @@
...
+    """ Operator Converted for converting Caffe ops to Relay ops """
+    def __init__(self, init_layer_dict, predict_layer, exp_tab):
+        self.init_layer_dict = init_layer_dict
+        self.predict_layer = predict_layer
+        self.exp_tab = exp_tab
+        self.new_bn = {}
+        self.changed_layers = None
+
+        self.convert_map = {
+            'BatchNorm': self.bn,
```
[the remark this answers is not in the document]
Comment (fernchen) [reply python/tvm/relay/frontend/caffe.py:46]: Good idea! I will do some changes

**Comment by fernchen** [comment]
Hi @tqchen @FrozenGene:
I adopted the suggestion that just add caffe env in the ci_cpu, see https://github.com/apache/incubator-tvm/pull/6023#discussion_r461209556
But now there are some errors in this pr when doing ci_gpu. I have read the error log, and find that it will try to generate docs by executing this script tutorials/frontend/from_caffe.py in tvmai/ci-gpu:v0.64, and obviously there is no Caffe env in tvmai/ci-gpu:v0.64. So is there any way to avoid this problem, since we only have caffe env in ci_cpu?

**Comment by FrozenGene** [comment]
> Hi @tqchen @FrozenGene:
> I adopted the suggestion that just add caffe env in the ci_cpu, see [#6023 (comment)](https://github.com/apache/incubator-tvm/pull/6023#discussion_r461209556)
> But now there are some errors in this pr when doing ci_gpu. I have read the error log, and find that it will try to generate docs by executing this script tutorials/frontend/from_caffe.py in tvmai/ci-gpu:v0.64, and obviously there is no Caffe env in tvmai/ci-gpu:v0.64. So is there any way to avoid this problem, since we only have caffe env in ci_cpu?

I think we have to install caffe to gpu docker too.

**Comment by fernchen** [comment]
> > Hi @tqchen @FrozenGene:
> > I adopted the suggestion that just add caffe env in the ci_cpu, see [#6023 (comment)](https://github.com/apache/incubator-tvm/pull/6023#discussion_r461209556)
> > But now there are some errors in this pr when doing ci_gpu. I have read the error log, and find that it will try to generate docs by executing this script tutorials/frontend/from_caffe.py in tvmai/ci-gpu:v0.64, and obviously there is no Caffe env in tvmai/ci-gpu:v0.64. So is there any way to avoid this problem, since we only have caffe env in ci_cpu?
> 
> I think we have to install caffe to gpu docker too.

Aha，what a bad news!  Shall we append caffe in gpu docker? I can proposal a new pr.

**Comment by fernchen** [comment]
> > > Hi @tqchen @FrozenGene:
> > > I adopted the suggestion that just add caffe env in the ci_cpu, see [#6023 (comment)](https://github.com/apache/incubator-tvm/pull/6023#discussion_r461209556)
> > > But now there are some errors in this pr when doing ci_gpu. I have read the error log, and find that it will try to generate docs by executing this script tutorials/frontend/from_caffe.py in tvmai/ci-gpu:v0.64, and obviously there is no Caffe env in tvmai/ci-gpu:v0.64. So is there any way to avoid this problem, since we only have caffe env in ci_cpu?
> > 
> > 
> > I think we have to install caffe to gpu docker too.
> 
> Aha，what a bad news! Shall we append caffe in gpu docker? I can proposal a new pr.

Need more advice from you @tqchen

**Comment by FrozenGene** [comment]
@fernchen please follow up.

**Comment by fernchen** [comment]
> @fernchen please follow up.

Thanks!
I have sent @tqchen  a email about how to deal with this problem, and wait for his responses. If there is no better resolution, I will prorpose a new PR and  add caffe environment to ci_gpu

**Comment by tqchen** [comment]
@fernchen let us skip the tutorials for now and only add the frontend cpu tests

**Comment by fernchen** [comment]
> @fernchen let us skip the tutorials for now and only add the frontend cpu tests

Thanks for your advice! I have modified the codes, please review the other part of them. @tqchen

**Comment by fernchen** [comment]
Hi @tqchen @FrozenGene :
Please let me know, if you have more suggestions about this pr.

**Comment by fernchen** [comment]
Hi @tqchen ：
I'm sorry to bother you, but I really need  more suggestions to push this pr forward.
Thanks very much!

**Comment by tqchen** [comment]
@fernchen please try to send another dummy commit to retrigger the CI, @FrozenGene please followup. I think we can merge once all the tests are resolved

**Comment by fernchen** [comment]
> @fernchen please try to send another dummy commit to retrigger the CI, @FrozenGene please followup. I think we can merge once all the tests are resolved

@tqchen Thanks for your suggestions! @FrozenGene I have send a new dummy commit for this pr and now all tests are passed. Please let me konw if you have more suggestions about this pr, thanks.

File: tests/python/frontend/caffe/test_forward.py [hunk 4]
```diff
@@ -0,0 +1,1003 @@
...
+        dtype_dict = {'data': 'float32'}
+
+    mod, params = relay.frontend.from_caffe(
+        init_net, predict_net, shape_dict, dtype_dict)
+
+    target = 'llvm'
+    target_host = 'llvm'
+
+    ctx = tvm.cpu(0)
+    with tvm.transform.PassContext(opt_level=2):
```
Comment (FrozenGene) [inline tests/python/frontend/caffe/test_forward.py:209]: Why can not be opt level 3?

File: tests/python/frontend/caffe/test_forward.py [hunk 5]
```diff
@@ -0,0 +1,1003 @@
...
+                          target_host=target_host,
+                          params=params)
+    dtype = 'float32'
+    m = graph_runtime.GraphModule(lib['default'](ctx))
+    if isinstance(data, (tuple, list)):
+        for idx, d in enumerate(data):
+            m.set_input('data' + str(idx), tvm.nd.array(d.astype(dtype)))
+    else:
+        m.set_input('data', tvm.nd.array(data.astype(dtype)))
+    m.set_input(**params)
```
Comment (FrozenGene) [inline tests/python/frontend/caffe/test_forward.py:221]: We don't need set parameter now in new module based runtime interface

File: tests/python/frontend/caffe/test_forward.py [hunk 6]
```diff
@@ -0,0 +1,1003 @@
...
+                      pad_w=2,
+                      kernel_h=3,
+                      kernel_w=5,
+                      stride_h=2,
+                      stride_w=1,
+                      dilation=[1, 2],
+                      weight_filler=dict(type="xavier"),
+                      bias_filler=dict(type="xavier"))
+
+    print("Testing layer Convolution pass!")
```
Comment (FrozenGene) [inline tests/python/frontend/caffe/test_forward.py:384]: remove all related print as we have assert

File: tests/python/frontend/caffe/test_forward.py [hunk 7]
```diff
@@ -0,0 +1,1003 @@
...
+             moving_average_fraction=moving_average_fraction,
+             eps=eps)
+
+
+def test_forward_BatchNorm():
+    """ BatchNorm """
+    data = np.random.rand(1, 3, 10, 10).astype(np.float32)
+    _test_batchnorm(data)
+    _test_batchnorm(data, moving_average_fraction=0.88, eps=1e-4)
+    print("Testing layer BatchNorm pass!")
```
Comment (FrozenGene) [inline tests/python/frontend/caffe/test_forward.py:296]: ditto

**Review by FrozenGene — CHANGES_REQUESTED** [review]

**Comment by FrozenGene** [comment]
> ```python
> from tvm.relay.frontend import caffe_pb2 as pb
> ```

where is from `tvm.relay.frontend import caffe_pb2 as pb`, i don't see it in this pr. And another thing is we must at least provide one instruction how to build own `caffe.proto` to related stuff.

File: tests/python/frontend/caffe/test_forward.py (hunk 4 above)
Comment (fernchen) [reply tests/python/frontend/caffe/test_forward.py:209]: > Why can not be opt level 3?

Sorry，this is a historical issue. In previous version, it will throw errors while building LRN layer(Alexnet) with opt_level 3. And this problem has been solved, so i will reset the opt_level to 3. Thanks

File: tests/python/frontend/caffe/test_forward.py (hunk 5 above)
Comment (fernchen) [reply tests/python/frontend/caffe/test_forward.py:221]: > We don't need set parameter now in new module based runtime interface

Thanks! I will delete this line.

**Comment by fernchen** [comment]
> > ```python
> > from tvm.relay.frontend import caffe_pb2 as pb
> > ```
> 
> where is from `tvm.relay.frontend import caffe_pb2 as pb`, i don't see it in this pr. And another thing is we must at least provide one instruction how to build own `caffe.proto` to related stuff.

Based on disccussion on https://github.com/apache/incubator-tvm/pull/6023#issuecomment-660408499, we will support BVLC Caffe in TVM firstly. So there is no need to compile own caffe.proto in this pr, we use the caffe.proto as follow:
```python
from caffe.proto import caffe_pb2 as pb
```

**Review by FrozenGene — APPROVED** [review]

**Comment by FrozenGene** [comment]
Thanks @fernchen It is merged now.
